### ISSUE TITLE
Status monitor inconsistent behaviour under load?

### ISSUE THREAD

**Issue by grke** [grke/burp#743]
Ziirish (burp-ui) says that he tried multiple 'burp -a m' processes, and starting with 4, he gets inconsistent results. Here are his logs, version 2.1.32:

```
--------------------------------------------------------------------------------                                                              [80/4633]
INFO in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:254]:
1 - query: 'c:cobalt' (cache: False)
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
DEBUG in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:275]:
1 => {'clients': [{'name': 'cobalt', 'labels': ['plip', 'plop', 'plup'], 'run_status': 'idle', 'protocol': 1, 'backups': [{'number': 918, 'timestamp':
1484601602, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}]}]}
--------------------------------------------------------------------------------
--
--------------------------------------------------------------------------------
INFO in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:254]:
2 - query: 'c:cobalt' (cache: False)
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
DEBUG in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:275]:
2 => {'clients': [{'name': 'cobalt', 'labels': ['plip', 'plop', 'plup'], 'run_status': 'idle', 'protocol': 1, 'backups': [{'number': 926, 'timestamp':
1485686402, 'flags': ['deletable', 'current', 'manifest'], 'logs': {'list': ['backup', 'restore', 'backup_stats', 'restore_stats']}}, {'number': 925, $
timestamp': 1485464400, 'flags': ['hardlinked', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 924, 'timestamp': 1485285602, '$
lags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 923, 'timestamp': 1485093604, 'flags': ['manifest'], 'logs': {'list': [$
backup', 'backup_stats']}}, {'number': 922, 'timestamp': 1485016802, 'flags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': $
21, 'timestamp': 1484898002, 'flags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 920, 'timestamp': 1484808002, 'flags': [$
deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 918, 'timestamp': 1484601602, 'flags': ['hardlinked', 'deletable', $
manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 911, 'timestamp': 1483474821, 'flags': ['hardlinked', 'deletable', 'manifest'], $
logs': {'list': ['backup', 'backup_stats']}}, {'number': 904, 'timestamp': 1482411601, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'lis$
': ['backup', 'backup_stats']}}, {'number': 876, 'timestamp': 1478431202, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup'$
 'backup_stats']}}, {'number': 848, 'timestamp': 1474108802, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_sta$
s']}}, {'number': 820, 'timestamp': 1471380002, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'num$
er': 792, 'timestamp': 1466715603, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}]}]}
--------------------------------------------------------------------------------
--
--------------------------------------------------------------------------------
INFO in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:254]:
1 - query: 'c:cobalt' (cache: False)
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
DEBUG in utils.status [/opt/workspace/burp-ui/burpui/../burpui/misc/backend/burp/utils.py:275]:
1 => {'clients': [{'name': 'cobalt', 'labels': ['plip', 'plop', 'plup'], 'run_status': 'idle', 'protocol': 1, 'backups': [{'number': 926, 'timestamp':
1485686402, 'flags': ['deletable', 'current', 'manifest'], 'logs': {'list': ['backup', 'restore', 'backup_stats', 'restore_stats']}}, {'number': 925, '
timestamp': 1485464400, 'flags': ['hardlinked', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 924, 'timestamp': 1485285602, 'f
lags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 923, 'timestamp': 1485093604, 'flags': ['manifest'], 'logs': {'list': ['
backup', 'backup_stats']}}, {'number': 922, 'timestamp': 1485016802, 'flags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 9
21, 'timestamp': 1484898002, 'flags': ['manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 920, 'timestamp': 1484808002, 'flags': ['
deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 918, 'timestamp': 1484601602, 'flags': ['hardlinked', 'deletable', '
manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'number': 911, 'timestamp': 1483474821, 'flags': ['hardlinked', 'deletable', 'manifest'], '
logs': {'list': ['backup', 'backup_stats']}}, {'number': 904, 'timestamp': 1482411601, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list
': ['backup', 'backup_stats']}}, {'number': 876, 'timestamp': 1478431202, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup',
 'backup_stats']}}, {'number': 848, 'timestamp': 1474108802, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stat
s']}}, {'number': 820, 'timestamp': 1471380002, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}, {'numb
er': 792, 'timestamp': 1466715603, 'flags': ['hardlinked', 'deletable', 'manifest'], 'logs': {'list': ['backup', 'backup_stats']}}]}]}
--------------------------------------------------------------------------------
--

The number before "query" is the process ID in the Queue.
So as you can see, the first 'c:cobalt' query does not return the full list of
backups if we compare it to the second or third queries results.
```

**Comment by grke** [comment]
Hello,

Thanks for the examples.

I think it is probably a burp thing.
Reading back your original report, if I understand correctly, it is always the first query of a 'burp -a m' process that might return less detail.

The burp server forks a new status monitor server side process each time a 'burp -a m' connects, and these server side processes themselves collect information from pipes connected to the burp master process (and probably from reading the disk too, if I remember correctly).
At the time of the first request/response, the server side processes might not yet have all the detailed information, and so just give the brief detail instead.

I have been thinking about how to make this better. I am coming to the conclusion that having all these pipes being written/read can probably be improved by getting rid of some of them. Probably things would be better if I used shared memory instead.
I have also been considering whether it would be appropriate to have only one forked monitor server side process instead of one for each 'burp -a m'. It would make I/O and permissions harder, so not 100% sure yet. Hmm, maybe it should fork a single process to read the disk info, and have multiple child processes share its memory.

So, I think the solution for this problem is some burp re-architecture. Also, I have never done anything with shared memory, so it should be interesting for me to do!

**Comment by ziirish** [comment]
I ran some other tests and indeed, they tend to show the problem only occurs an the first queries.

Besides, the PoC above is not 100% representative of the actual workload when the issue occurs.

I can reproduce the issue 100% of the time with the complete burp-ui code when computing stats of a client so I have to run the following queries: c:, c:client:b:1, c:client:b:2, ... c:client:b:1:l:backup_stats, etc.

What bothers me though is the problem occurs 100% of the time when I run more concurrent `burp -a m` processes than the number of available cores

**Comment by grke** [comment]
Yes, it's just that when the system is under load, it is not able to load the data fast enough in order to have it all ready for the first response, and it is most likely the data that comes from disk reads that is most affected.

If there was just one burp process that dealt with the disk reads, then it would be more consistent because each 'burp -a m' would return the same thing. And there would be fewer disk reads, of course.

The problem that I am trying to solve right now (conceptually, as I have coded nothing yet), is that the shared memory that mmap() gives you cannot be resized unless you are using Linux (which has mremap(). Which means you have to allocate it up front.

Maybe the disk read process can check how many files are in clientconfdir when it starts, and make a sensible guess. Then if it needs more, and it is Linux, it can mremap(), otherwise it just has to exit (which will make all the server side child status processes also exit). It will need a config item (or items - memory for clients / memory for backups per client) to set so that people can have it really big if they like.

**Comment by grke** [comment]
I suspect the problem will still occur with your idea. If it doesn't, it is probably just deferred until you have even more processes.

**Comment by ziirish** [comment]
Sure, no problem. Thanks for the efforts you put into this as usual.

**Comment by ziirish** [comment]
Hello,

I managed to have a PoC able to reproduce the issue.

Here is a example output:

```
[...snip...]
got 2 different results for 'testclient'
4 times:                   
[                                         
    {
        "number": 904,
        "deletable": true,
        "date": 1482411601,
        "encrypted": false,
        "received": 20353330,
        "size": 63426095044,
        "end": 1482412062
    }
]
6 times:
[
    {
        "number": 792,
        "deletable": true,
        "date": 1466712001,
        "encrypted": false,
        "received": 134828302,
        "size": 49779850666,
        "end": 1466712573
    },
[...snip...]
    {
        "number": 925,
        "deletable": false,
        "date": 1485464400,
        "encrypted": false,
        "received": 6805919,
        "size": 73114364390,
        "end": 1485465131
    },
    {
        "number": 926,
        "deletable": true,
        "date": 1485686401,
        "encrypted": false,
        "received": 11742702,
        "size": 73110778353,
        "end": 1485686998
    }
]
```


The main reproduction factor is the client concurrency level, it means I'm able to reproduce the issue even with only one `burp -a m`  process.

I'm still not sure to understand this behavior though because my monitor pool is supposed to wait until a process is available so even if the queries are scheduled asynchronously, they are processed one at the time because there is only one worker available.
But I understand this must push high pressure on the burp-server (or status monitor child process)

Here is the whole locking mechanism:

```python
    @asynccontextmanager                                                                 
    async def get_mon(self, ident) -> Monitor:                                           
        self.logger.info(f'{ident} - Waiting for a monitor...')                          
        t1 = trio.current_time()                                                         
        mon = await self.monitor_pool.get()  # type: Monitor                             
        t2 = trio.current_time()                                                         
        t = t2 - t1                                                                      
        self.logger.info(f'{ident} - Waited {t:.3f}s')                                   
        yield mon                                                                        
        self.logger.info(f'{ident} - Releasing monitor')                                 
        await self.monitor_pool.put(mon)
```

And is used like this:

```python
async with self.get_mon(ident) as mon:                           
    response = mon.status(query, timeout=self.timeout, cache=False, raw=True)
```

I know you are not quite familiar with python, but the idea is we have a kind of shared structure between our "threads" (these are not real threads, but they kind of work like threads): the `self.monitor_pool` object (it's a shared queue).
We acquire a `Monitor` object (this is the `burp -a m` python interface I wrote) through a call to the `self.get_mon` function, we "return" it to the caller function through the `yield` statement (it's a kind of re-entrant `return`).
At the end of the `async with` block, we return back to the `yield` in the `get_mon` func so we can put the `Monitor` object back to our `monitor_pool`.

This is to say, when our shared queue is empty, the `get_mon` function blocks until we put back a `Monitor` in it.


Attach are the script used to reproduce the issue with their usage:

server:

```
python monitor.py -b 127.0.0.1 -p 11112 -C 1 -c /tmp/burp/burp.conf -vvvv
```

With `-c` the burp client config file, `-C` the number of `burp -a m` processes to launch and `-vvvv` the log verbosity (you can decrease it for fewer logs)

client:

```
python client.py -H 127.0.0.1 -p 11112 -r 10 -c 1 testclient
```

With `-r` the number of times to repeat the requests, `-c` the concurrency level (ie. the number of queries to run concurrently client side) and `testclient` the client you want to compute stats for.

[client.txt](https://github.com/grke/burp/files/2303739/client.txt)
[monitor.txt](https://github.com/grke/burp/files/2303740/monitor.txt)

**Comment by grke** [comment]
It is strange that it is giving you some random backup number in the middle of the list (904) and not the latest one. I don't understand that.

**Comment by grke** [comment]
I will try to run youe scripts when I have a chance. Thank you.

How many backups does testclient have in total? Do they all have minimal manifests?

**Comment by ziirish** [comment]
I can reproduce this behavior with other clients as well.
They have between 10 and 14 backups.

But none of the backups were made with the 2.1.32 server they are old backups (some were even made with 1.4.40 server).

Anyway, digging a bit into the logs I got, I noticed this:

```
--------------------------------------------------------------------------------                                                                                                                                                               
INFO in monitor.status [monitor.py:293]:                                                                                                                                                                                                       
1 - query: 'c:testclient' (cache: False, raw: True)                                                                                                                                                                                                
--------------------------------------------------------------------------------                                                                                                                                                               
--------------------------------------------------------------------------------                                                                                                                                                               
DEBUG in monitor.status [monitor.py:320]:                                                                                                                                                                                                      
1 => {"clients":[{"name":"testclient","labels":["plip","plop","plup"],"run_status":"idle","protocol":1,"backups":[{"number":926,"timestamp":1485686402,"flags":["deletable","current","manifest"],"logs":{"list":["backup","restore",              "backup_stats","restore_stats"]}},{"number":925,"timestamp":1485464400,"flags":["hardlinked","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":924,"timestamp":1485285602,"flags":["manifest"],"logs":{"list":["backup",        "backup_stats"]}},{"number":923,"timestamp":1485093604,"flags":["manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":922,"timestamp":1485016802,"flags":["manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":921,     "timestamp":1484898002,"flags":["manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":920,"timestamp":1484808002,"flags":["deletable","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":918,"timestamp":1484601602,  "flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":911,"timestamp":1483474821,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":904,          "timestamp":1482411601,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":876,"timestamp":1478431202,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}}, {"number":848,"timestamp":1474108802,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}},{"number":820,"timestamp":1471380002,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup",     "backup_stats"]}},{"number":792,"timestamp":1466715603,"flags":["hardlinked","deletable","manifest"],"logs":{"list":["backup","backup_stats"]}}]}]}                                                                                            
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------                                                                                                                                                               
INFO in monitor.status [monitor.py:293]:
1 - query: 'c:testclient:b:924' (cache: False, raw: True)                                                                                                                                                                                          
--------------------------------------------------------------------------------                                                                                                                                                               
--------------------------------------------------------------------------------                                                                                                                                                               
DEBUG in monitor.status [monitor.py:320]:                                                                                                                                                                                                      
1 => {"clients":[{"name":"testclient","labels":["plip","plop","plup"],"run_status":"idle","protocol":1,"backups":[{"number":924,"timestamp":1485285602,"flags":["manifest"],"logs":{"list":["backup","backup_stats"]}}]}]}                         
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------                                                                                                                                                               
INFO in monitor.status [monitor.py:293]:                                                                                                                                                                                                       
1 - query: 'c:testclient' (cache: False, raw: True)                                                                                                                                                                                                
--------------------------------------------------------------------------------                                                                                                                                                               
--------------------------------------------------------------------------------                                                                                                                                                               
DEBUG in monitor.status [monitor.py:320]:                                                                                                                                                                                                      
1 => {"clients":[{"name":"testclient","labels":["plip","plop","plup"],"run_status":"idle","protocol":1,"backups":[{"number":924,"timestamp":1485285602,"flags":["manifest"],"logs":{"list":["backup","backup_stats"]}}]}]}                         
--------------------------------------------------------------------------------
```

So the first query `c:cobalt` return the expected results, then we start iterating over the backups hence the `c:testclient:b:924` query at the same time, another `c:cobalt` request is sent.
The `1` is the monitor ID so we are working with the same `burp -a m` process.

Here are some possible explanations:

1. the burp server monitor has some caching/memory issues resulting in the wrong results when we push too much pressure on it
2. the burp client process has some caching/memory issues resulting in the wrong output when we push too much pressure on it
3. burp-ui has some caching/memory issues resulting in the wrong output when we push too much pressure on it

I have disabled the cache as the logs report, so the queries are well sent to the monitor. But maybe things start to mess up when I read the monitor stdout and I read twice the same thing because the server did not have time to answer yet (due to the high pressure everywhere).

One way to prevent this would be to add an end-delimiter when the server finished to answer.

Something like `END:<query>` would ensure me I read the right output.

**Comment by grke** [comment]
Sorry, I have struggled to make pyenv do anything useful.

```
graham@small:~$ pyenv versions
  system
* 3.6.6 (set by /home/graham/.pyenv/version)
graham@small:~$ python --version
Python 2.7.13
```

Trying to 'pip install trio' just gives some errors, presumably because it is not python 3.6.x.
And, the same for your monitor.txt script above.

**Comment by ziirish** [comment]
I know the '}' is easily parseable and have no problem to parse the status monitor output.
But I have no way to know what 'block' answers which 'query' and I suspect that's probably what is the source of this inconsistency issue.

-----

you need to enable your env: `pyenv shell 3.6.6`

**Comment by ziirish** [comment]
I'm available on [IRC](https://webchat.freenode.net/) in the `#burp-ui` channel if you need some "interactive" help.

**Comment by grke** [comment]
OK, thanks,I have managed to run monitor.py.

**Comment by ziirish** [comment]
I tried to duplicate the burp `directory` tree in `/dev/shm` but the problem still occurs on my development VM.

Then I tried to reproduce the issue on another VM with 8 vCPUs but so far, I was unable to reproduce it (just like on your workstation).

**Comment by grke** [comment]
I have a question about what you are doing:

Do you intend, for example, to start 2 'burp -a m' processes, and then choose one at random for a query?
Like:

```
Process 1: c:testclient
Process 2: c:otherclient
Process 1: c:otherclient:b:85
Process 2: c:testclient:b:38
Process 1: c:testclient:b:38:p:/path
Process 2: c:otherclient:b:85:p:/path
```

Or, would queries for a particular client stick to one particular 'burp -a -m' process?
Like:

```
Process 1: c:testclient
Process 2: c:otherclient
Process 1: c:testclient:b:38
Process 2: c:otherclient:b:85
Process 1: c:testclient:b:38:p:/path
Process 2: c:otherclient:b:85:p:/path
```

**Comment by grke** [comment]
^ Currently, the random option would mean that the server side child process for process 1 would load the data separately to that for process 2. That is, it would be loaded twice, because the server side child processes do not share memory.
Currently, I have not got a really good solution idea for this - I cannot use my shared memory plan for browsing the contents of backups because I do not know in advance how much memory will be needed for the directory tree, and only Linux appears able to resize shared memory with mremap(). One option is to make it a Linux-only feature, I guess.

**Comment by grke** [comment]
^ It doesn't matter about the burp client-side, I am talking Linux-only-feature server side.

**Comment by ziirish** [comment]
The problem also occurs when there is only one status monitor process (ie. `-C 1` with monitor.py)

**Comment by ziirish** [comment]
I understand you want to get rid of some of the pipes between the main burp server process and the status monitor child, but wouldn't the shared memory thing involve too much headache?

I mean, what if two child processes receive the same query at the same time? They will both need the same data in the end but since there is no such available data in the cache they will compute them independently, then they will need to acquire a lock to update the shared memory and finally one of them will realize it did all this for nothing.

Also I'm not sure the *BSD guys will like a Linux only status monitor.

Instead you could let each status monitor child handle its file reading. On Unix there is no problem to access the same file concurrently in read-only. Besides, I'm pretty sure the kernel has some magic tricks to optimize those concurrent read operations.
The downside is the more status child you have, the more memory you will consume.

If you really need some shared memory, then using an external cache product is maybe easier to implement (something like memcached or redis).

**Comment by grke** [comment]
My idea was for a separate single process to handle the disk reads, and arrange it nicely in memory.
Then the child processes would just traverse the in-memory structure. The structure would likely be implemented in such a way so as not to need locks. 
Currently, the child processes do read and compute the data independently, both from pipes, and from the disk.
I think that is OK for the client and backup list, but I cannot think of a way to make it work for backup contents without needing mremap(), or some configurable max shared memory size (ie, non-Linux would be forced to set the max shared memory size, and exceeding it would cause things to log errors and exit).

Your idea of redis or memcached - that is an interesting suggestion that I will now think about!
I use both at work, but never considered them for burp.
Thanks!

**Comment by ziirish** [comment]
Since I recommend redis when deploying burp-ui in a "production" environment, I suppose it is acceptable to implement its support for the status monitor as long as its stay an option (I like how KISS burp is)

**Comment by grke** [comment]
Don't worry, I am just trying to think of the best plan at the moment, not set on anything.

**Comment by ziirish** [comment]
Awesome, thanks!

I'll give it a try as soon as I can and let you know about my consistency results.

**Comment by grke** [comment]
To clarify, with the last post above, you have three request/responses.
Are each of them using a different 'burp -a m'? Or are they all three using the same 'burp -a m'?

**Comment by grke** [comment]
OK, thanks for the explanation, this is very interesting evidence, I have a suspicion about a particular piece of code. I will investigate when I have a moment.

**Comment by grke** [comment]
Yeah, I think we've found the problem! Patch coming soon.

**Comment by grke** [comment]
Patch in master now.
Please let me know if you can no longer reproduce the problem, then we can close this issue.

**Comment by ziirish** [comment]
Awesome, I haven't been able to reproduce the issue so far!

Could you please briefly explain what was the problem? I'm not sure to understand your patch.

Thanks

**Comment by grke** [comment]
The clue that led me to realising what was going on was that it was a single 'burp -a m' process, and (3) repeating the same response as (2). Since the response markers were given, I was able to pinpoint the section of code that was doing something weird. Then I realised that this section of code had no utest coverage, so I wrote a utest that duplicated the problem. Then I fixed the code.

The previous get_str() function was broken, as it would read on past the end of the command that it had been asked to parse.
So, in the third case above, it could have this in memory (dependent on randomness):
c:cobalt(null)b:911(null)
And it would be treated as:
c:cobalt:b:911(null)
The fix was to use strtok to tokenise the command instead of trying to do it manually.

Thanks for reporting this! Now closing.
